### DISCUSSION TITLE
How to change database URL via an environment variable?

### DISCUSSION THREAD

**Discussion by bubelov** [rwf2/Rocket#1805] [Questions] [answered]
Let's say I have the following configuration in `Rocket.toml`

```toml
[default.databases.main]
url = "data.db"
```

How do I change it via an environment variable? I tried this (with no luck):

```bash
ROCKET_DATABASES_MAIN_URL=newdb cargo run
```

Indeed, database URL is just an example, my real use case is adding sensitive data such as API keys, which shouldn't be laying in `Rocket.toml`. I saw plenty of issues related to configuration system being too rigid (#1233, #852, etc), but they are now closed so I assume it's supposed to be working in 0.5?

**Comment by jebrosen** [comment, marked as answer]
There are a few examples of configuration via environment variables in the crate-level documentation for [`rocket_sync_db_pools`](https://api.rocket.rs/master/rocket_sync_db_pools/index.html#environment-variables) and [`rocket_db_pools`](https://api.rocket.rs/master/rocket_db_pools/index.html#configuration), both of which will be in the 0.5 release.

(It happens to be identical for 0.4: [`rocket_contrib::databases`](https://api.rocket.rs/v0.4/rocket_contrib/databases/index.html#environment-variables)).

**Reply by bubelov** [reply]
@jebrosen I saw you recommending this approach here https://github.com/SergioBenitez/Rocket/issues/1233#issuecomment-588401909. It's a very nice trick, but the linked issue implies that the system has been revamped, addressing all of those use cases in a clean way. I'll stick to this approach for now since it works and it's documented, thanks!

**Reply by jebrosen** [reply]
Ah! For nesting in particular, it does not split by `_` by default. But this could be done through [custom configuration](https://rocket.rs/v0.5-rc/guide/configuration/#custom-providers) with this method: [`figment::providers::Env.split()`](https://docs.rs/figment/0.10.6/figment/providers/struct.Env.html#method.split).

Rocket also accepts the environment variable `ROCKET_databases.main.url`. Bash (and likely many other programs) doesn't like this variable name, but the `env` command can do it on Linux: `env ROCKET_databases.main.url=newdb cargo run`
